components/homme/src/share/compose/compose_homme.cpp

```diff
@@ -31,46 +30,109 @@ void TracerArrays<MT>::alloc_if_not () {
 #endif
 
 template <typename MT>
-void sl_h2d (TracerArrays<MT>& ta, bool transfer, Cartesian3D* dep_points) {
+void sl_traj_h2d (TracerArrays<MT>& ta, Real* dep_points, Real* vnode,
+                  Real* vdep, Int ndim) {
 #if defined COMPOSE_PORT
+# if defined COMPOSE_HORIZ_OPENMP
+# pragma omp master
+  {
+# endif
   ko::fence();
   ta.alloc_if_not();
   const Int nelemd = ta.nelemd, qsize = ta.qsize, np2 = ta.np2, nlev = ta.nlev;
-  const DepPointsH<MT> cart_h(reinterpret_cast<Real*>(dep_points), nelemd, nlev, np2);
-  const auto dep_points_h = ko::create_mirror_view(ta.dep_points);
-  for (Int ie = 0; ie < nelemd; ++ie)
-    for (Int lev = 0; lev < nlev; ++lev)
-      for (Int k = 0; k < np2; ++k)
-        for (Int d = 0; d < 3; ++d)
-          dep_points_h(ie,lev,k,d) = cart_h(ie,lev,k,d);
-  ko::deep_copy(ta.dep_points, dep_points_h);
-  if ( ! transfer) return;
-  const auto qdp_m = ko::create_mirror_view(ta.qdp);
-  const auto dp_m = ko::create_mirror_view(ta.dp);
+  const DepPointsH<MT> cart_h(dep_points, nelemd, nlev, np2, ndim);
+  ko::deep_copy(ta.dep_points, cart_h);
+  if (vnode) {
+    const DepPointsH<MT> h(vnode, nelemd, nlev, np2, ndim);
+    ko::deep_copy(ta.vnode, h);
+  }
+  if (vdep) {
+    const DepPointsH<MT> h(vdep, nelemd, nlev, np2, ndim);
+    ko::deep_copy(ta.vdep, h);    
+  }
+# ifdef COMPOSE_HORIZ_OPENMP
+  }
+# pragma omp barrier
+# endif
+#endif
+}
+
+template <typename MT>
+void sl_traj_d2h (const TracerArrays<MT>& ta, Real* dep_points, Real* vnode,
+                  Real* vdep, Int ndim) {
+#if defined COMPOSE_PORT
+# if defined COMPOSE_HORIZ_OPENMP
+# pragma omp master
+  {
+# endif
+  ko::fence();
   const auto q_m = ko::create_mirror_view(ta.q);
-  for (Int ie = 0; ie < nelemd; ++ie)
-    for (Int iq = 0; iq < qsize; ++iq)
+  const Int nelemd = ta.nelemd, np2 = ta.np2, nlev = ta.nlev;
+  const DepPointsH<MT> dep_points_h(dep_points, nelemd, nlev, np2, ndim);
+  ko::deep_copy(dep_points_h, ta.dep_points);
+  if (vnode) {
+    const DepPointsH<MT> h(vnode, nelemd, nlev, np2, ndim);
+    ko::deep_copy(h, ta.vnode);
+  }
+  if (vdep) {
+    const DepPointsH<MT> h(vdep, nelemd, nlev, np2, ndim);
+    ko::deep_copy(h, ta.vdep);    
+  }
+# ifdef COMPOSE_HORIZ_OPENMP
+  }
+# pragma omp barrier
+# endif
+#endif  
+}
+
+template <typename MT>
+void sl_h2d (TracerArrays<MT>& ta, bool transfer, Real* dep_points, Int ndim) {
+#if defined COMPOSE_PORT
+# if defined COMPOSE_HORIZ_OPENMP
```

Comment:
If COMPOSE_PORT is defined, then we're building for the kokkos dycore, right? In that case we should not have horiz openmp, right? I thought horiz openmp was used in the f90 dycore only...

Reply:
Same in a few other places.

Reply:
The Kokkos code path is supported in the F90 dycore; it's just not used in practice because of array ordering (requiring transposes and extra memory) and needing to support HORIZ_OPENMP on CPUs.

Reply:
but the cmake logic seems to be such that COMPOSE_PORT is only true for the compose_f90 lib, which is the only one linked against the f90 targets. Maybe I'm misreading the cmake logic though...

Reply:
(For the composec++ lib, not the composef90 one, but I get your point.) Yes, that's right, because in practice we don't want to use COMPOSE_PORT for the F90 dycore. That is, nobody would purposely build a dycore that way.

But it does actually work. For such a niche thing, I haven't wanted to waste testing resources to keep it working, but I use the capability when doing dev work on occasion. In short, I do want to retain lines of the sort you're highlighting even though our automated builds never activate them. Note the lines you highlighted correspond to a bridge routine to do the transposes. (Detail: The bridge routine is also used in compose_ut unit tests for F90-vs-C++ comparisons, but of course in this case HORIZ_OPENMP is, as you pointed out, always false.)